README.md
# wsclient-evented
> A (very) simple WebSocket client wrapper written in ES2015, coupled with an event emitter.

## Why should I use it?
wsclient-evented is just an easy wrapper for the native WebSocket client.
It adds some often needed functionality like

- [x] Relative small in size (8.5kb minified)
- [x] Automatic reconnects when connections get closed
- [x] Send method automatically opens the connection if it is closed
- [x] Possibility to listen for any server events without big switch statements in onMessage via facebooks fbemitter
- [x] Automatically parsing of payload data (when using the default message handler)
- [x] Cross browser compatible* (at least with all browsers that support WebSockets :))

---

## Missing features
- [ ] Binary data support has not been implemented (in the default listener only!)

---

## Compatibility
- [x] Google Chrome
- [x] Apple Safari
- [x] Mozilla Firefox
- [x] Microsoft Edge
- [x] Microsoft Internet Explorer (10, 11)

---

## Installation
There are two possibilities to use the client:

#### Old school:
Just add the file dist/browser.js into your url to make window.WsClientEvented available.

Example:
```html
<script src="wsclient-evented/dist/wsclientevented.js"></script>
<script>
var wsClient = new WsClientEvented('ws://url');
wsClient.send('x');
</script>
```

---

#### ES2015 import
If you are using webpack or native ES2015 imports, you may use the client like this:

```javascript
import WsClient from 'wsclient-evented';
let wsClient = new WsClient('ws://url');
wsClient.send('x');
```

---

### Examples
You can see a working example in ```lib/index.html```.
This example connects to a node.js WebSocket server that is configured to listen for the configured message format. Just use ```npm run test:watch``` to see the connect and event firing happen via your favorite webdeveloper console tool.

---

### Usage
The constructor expects the following arguments:

#### url (String, required):
The url to connect to as defined in the ws protocol (type://host:port). This must always be provided.

#### protocols (String, Array or null, optional):
One or more protocols to use as described in the WebSocket spec. Automatically set to ```null``` if no protocol should be used.

#### options (Object, optional):
Key/Value option object, used to override the default settings.
wsclient-evented can be configured with an options array on initialisation.
Provided options will automatically override the defaults shown below.
The following options are available:

```javascript
bool autoOpen [default: true] Automatically connect the WebSocket on initialisation?
bool autoReconnect [default: true] Automatically reconnect the WebSocket if connection is lost?
bool forceCloseOnReload [default: false] Send a forced close on browser reload? **Only available when using the client in the main thread!**
bool debug [default: false] Show debug output?
int reconnectInterval [default: 1000] Delay in ms for reconnect tries
float reconnectDecay [default 1.5] Factor for reconnects
int timeoutInterval [default: 2000] Time in ms to wait for connects
string binaryType [default: blob] Binary type for websocket transmission
string payloadItemName [default: "payload"] The name of the payload key, used for send
int maxReconnects [default: null] Maximal amount of reconnects. Set to null for inifinite
int maxReconnectTimeout [default: 5000] Maximal amount of milliseconds to wait for reconnects
int maxSendTries [default: 10] Maximal amount of tries for ws send to fail for the same request
function onBeforeWsOpen [default: null] Called before WebSocket connections are established
function onWsOpen [default: null] Called when a WebSocket connection becomes ready
function onWsMessage [default: defaultMessageHandler] Called on all WebSocket onMessage events. Defaults to emitting via fbemitter
function onWsClose [default: null] Called when a WebSocket connection is closed
function onWsError [default: null] Called when a WebSocket connection throws errors
function onMaxReconnects [default: null] Called when the maximal amount of reconnects is reached
function onWsTimeout [default: null] Called when a WebSocket connection times out
```

---

### Methods
The following methods are available for usage after a new wsclient-evented was created.

#### addListener(string evt, function callBack):
Adds an event listener for event ```evt``` that will trigger the callback ```callBack```.
Events registered this way will fire when a message with the given type was received via the WebSocket.

```javascript
// assuming ws is a valid wsclient-evented instance and test-response is an event that got send from the server
ws.addListener('test-response', function(eventType, wsEvent, data) {
  console.log('got new event of type ' + eventType);
  console.log('original websocket event: ', wsEvent);
  console.log('received data: ', data);
});
```

---

#### removeListener(string evt):
Removes all events named ```evt``` from the WebSockets onMessage listener.

---

#### send(string type, object payload[optional]):
Sends a message of type ```type``` with a payload of ```payload```.
Payload may be any json serializeable object or omitted to send an empty message.
This method will automatically try to reconnect the WebSocket if it is not connected before sending.
The amount of send tries can be configured via ```options.maxSendTries```.

---

##### wsclient-evented websocket message format:
The client sends a special format when using text type messages.
It is currently defined like this for the default message handler:
```javascript
{
  type: type,
  payload: payload
}
```

This format will get stringified before sending and ***must*** therefore be translated on the server. Server events, must send data in the same way!
See ```karma.conf.js``` for an example how to do this.

```javascript
// ... snip
connection.on('message', (msg) => {

  if(msg.type === 'utf8') {

    let payload = JSON.parse(msg.utf8Data);

    switch(typeof payload) {
      case 'string':
        connection.sendUTF(JSON.stringify(`${payload} - response`));
        break;

      case 'object':
      default:
        connection.sendUTF(JSON.stringify({
          type: `${payload.type}-response`,
          payload: payload.payload
        }));
    }
  }
});
// ... snip
```

---

#### open(bool reconnecting[optional]):
The open function will open a new WebSocket connection to the specified host.
It also sets up listeners for opening/closing and timeouts.
Automatically called when ```options.autoOpen``` is set to true (the default).
The flag reconnecting indicates a automatic reconnection attempt if the WebSocket closes unexpectedly.

---

#### close(int code[optional], string reason[optional]):
Closes the WebSocket. Will also make sure the socket does not automatically reconnect if ```options.autoReconnect``` is set to true (the default).
After closing the WebSocket, you will have to manually call wsclient-evented#open to connect the socket again.

---

## Defining own message handlers

The provided default message handler will just allow json in a particular format to be used. It is defined like this:

```javascript
/**
 * Default regular emitter
 * @param {MessageEvent} e Original websocket event
 * @param {WsClientEvented} instance
 */
const defaultMessageHandler = (e, instance) => {
  console.log('called orig bound baby!', e, instance, 'fuck!');
  const data = JSON.parse(e.data);
  const responseType = typeof data;

  if (responseType === 'object' && data.type) {
    instance.emitter.emit(data.type, e, data);
  }
};
```

You may override it by adding your own implementation of `onWsMessage` as option when instanciating a new client:

```javascript
const instance = new WsClientEvented('ws://', null, {
  onWsMessage: (evt, instance) => {
    console.log(evt, instance);
    instance.emitter.emit(evt.MY_CUSTOM_PAYLOAD_NAME, evt.data);
  }
});
```

---

### Special thanks
This client would not have been possible without the work of [Joe Walnes](https://github.com/joewalnes) who build [reconnecting-websocket](https://github.com/joewalnes/reconnecting-websocket). Check out this project if you need a much lighter, lower level version of what I try to accomplish.

---

### Licence
[MIT license](http://opensource.org/licenses/MIT)


CHANGELOG.md
# 2.2.0
1. Upgraded `fbemitter` to latest version 3.0.0.

# 2.1.0
1. Upgraded dev packages to latest stable versions
2. Added new callback `onMaxReconnects` that fires after timeouts

# 2.0.0
1. Upgraded development toolchain to latest package versions (eslint, webpack, babel, karma).
2. Changed code to validate against airbnb coding guidelines.
3. Provided overrideable default `onWsMessage` handler with default fallback for version `1.x.x`.
4. Exchanged `window` keyword for `self` where appropriate to make it work in WebWorkers without breaking.
5. Export only needed files for install on `package.json`

# 1.2.0
1. Added support for forced closing of connections on browser reload (defaults to false)
2. Updated all build packages to latest versions.

# 1.1.0
1. Added support for dynamic payload keys (can be configured via settings.payloadItemName)

# 1.0.0
Initial release


cfg/test.js
var path = require('path');

module.exports = {
  devtool: 'eval',
  module: {
    loaders: [
      {
        test: /\.js$/,
        loader: 'babel-loader',
        include: [
          path.join(__dirname, '/../lib'),
          path.join(__dirname, '/../test')
        ]
      },
      {
        test: /\.js$/,
        loader: 'babel-loader',
        include: [
          path.join(__dirname, '/../lib')
        ],
        loader: 'isparta'
      }
    ]
  }
};


test/loadtests.js
'use strict';

// Add support for all files in the test directory
const testsContext = require.context('.', true, /(Test\.js$)/);
testsContext.keys().forEach(testsContext);


test/lib/indexTest.js
/* eslint-env node, mocha */
/* global expect */
/* eslint no-console: 0 */
/* eslint no-unused-expressions: 0 */
import WsClientEvented from '../../lib/index';

const wsUrl = 'ws://localhost:9999';
const wsProtocol = 'wsclient-evented-protocol';

describe('when creating a new WsClientEvented instance', () => {

  it('should be able to run with just the url', () => {

    const instance = new WsClientEvented(wsUrl, '', {
      autoOpen: false,
      autoReconnect: false,
    });

    expect(instance.url).to.equal(wsUrl);
    expect(instance.protocols).to.be.null;
    expect(instance.settings).to.include({
      autoOpen: false,
      autoReconnect: false,
      debug: false,
      forceCloseOnReload: false,
      binaryType: 'blob',
      payloadItemName: 'payload',
      maxReconnects: null,
      maxSendTries: 10,
      maxReconnectTimeout: 5000,
      reconnectInterval: 1000,
      reconnectDecay: 1.5,
      timeoutInterval: 2000,
      onBeforeWsOpen: null,
      onWsOpen: null,
      onWsError: null,
      onWsClose: null,
      onWsTimeout: null,
    });
    expect(instance.settings.onWsMessage).to.be.a('function');
  });

  it('should be able to run with url and protocol', () => {
    const instance = new WsClientEvented(wsUrl, 'test');
    expect(instance.url).to.equal(wsUrl);
    expect(instance.protocols).to.deep.equal('test');
    instance.close();
  });

  it('should be able to override the default options with the provided ones', () => {
    const instance = new WsClientEvented(wsUrl, 'test', {
      debug: true,
      autoOpen: false,
      autoReconnect: false,
    });

    expect(instance.url).to.equal(wsUrl);
    expect(instance.protocols).to.equal('test');
    expect(instance.settings).to.include({
      autoOpen: false,
      autoReconnect: false,
      debug: true,
      forceCloseOnReload: false,
      binaryType: 'blob',
      payloadItemName: 'payload',
      maxReconnects: null,
      maxSendTries: 10,
      reconnectInterval: 1000,
      reconnectDecay: 1.5,
      timeoutInterval: 2000,
      maxReconnectTimeout: 5000,
      onBeforeWsOpen: null,
      onWsOpen: null,
      onWsError: null,
      onWsClose: null,
      onWsTimeout: null,
    });
    expect(instance.settings.onWsMessage).to.be.a('function');
    instance.close();
  });
});

describe('when opening a websocket', () => {

  let instance;
  beforeEach(() => {
    instance = new WsClientEvented(wsUrl, wsProtocol, {
      autoOpen: false,
      autoReconnect: false,
    });
  });

  afterEach(() => {
    instance.close();
  });

  it('should use a fluent interface', () => {
    expect(instance.open()).to.be.equal(instance);
  });

  it('should use auto opening per default', (done) => {
    const ws = new WsClientEvented(wsUrl, wsProtocol, {
      onBeforeWsOpen: (e, current) => {
        expect(current).to.be.instanceOf(WsClientEvented);
        done();
      },
    });

    expect(ws.settings.autoOpen).to.be.true;
    expect(ws.ws).to.be.null;
    ws.close();
  });

  it('should set the current websocket', () => {
    expect(instance.ws).to.be.null;
    instance.open();
    expect(instance.ws).to.be.instanceOf(WebSocket);
  });

  it('should close the old websocket and open a new one when called twice', () => {
    instance.open();
    const oldWs = instance.ws;

    instance.open();
    const newWs = instance.ws;

    expect(oldWs).to.be.instanceOf(WebSocket);
    expect(newWs).to.be.instanceOf(WebSocket);
    expect(newWs).to.not.equal(oldWs);
  });

  it('should listen for the onWsTimeout event if the connection is not possible', (done) => {
    new WsClientEvented('ws://localhost:9999', null, {
      debug: false,
      autoReconnect: false,
      timeoutInterval: 5,
      onWsTimeout: function(e, instance) {
        expect(instance).to.be.instanceOf(WsClientEvented);
        instance.close();
        done();
      }
    });
  });

  describe('when adding events at the constructor', () => {
    it('should fire the the onBeforeWsOpen event', (done) => {
      new WsClientEvented(wsUrl, wsProtocol, {
        autoReconnect: false,
        onBeforeWsOpen: (e, instance) => {
          expect(instance).to.be.instanceOf(WsClientEvented);
          instance.close();
          done();
        }
      });
    });

    it('should fire the the onWsOpen event', (done) => {
      new WsClientEvented(wsUrl, wsProtocol, {
        autoReconnect: false,
        onWsOpen: (e, instance) => {
          expect(e.type).to.equal('open');
          expect(e.target.readyState).to.equal(1);
          expect(instance).to.be.instanceOf(WsClientEvented);
          instance.close();
          done();
        }
      });
    });

    it('should fire the onWsMessage event', (done) => {
      new WsClientEvented(wsUrl, wsProtocol, {
        autoReconnect: false,
        onWsOpen: (e, instance) => {
          instance.ws.send(JSON.stringify('message'));
        },
        onWsMessage: () => {
          expect(true).to.be.true;
          instance.close();
          done();
        }
      });
    });

    it('should fire the the onWsClose event', (done) => {
      const instance = new WsClientEvented(wsUrl, wsProtocol, {
        autoReconnect: false,
        onWsClose: (e, self) => {
          expect(e.type).to.equal('close');
          expect(e.target.readyState).to.be.within(2, 3);
          expect(self).to.be.instanceOf(WsClientEvented);
          done();
        },
      });
      instance.ws.close();
    });

  });
});

describe('when closing a websocket connection', () => {

  it('should use a fluent interface', () => {
    const instance = new WsClientEvented(wsUrl, wsProtocol);
    expect(instance.close()).to.be.equal(instance);
  });

  it('should close the websocket with the default reason', (done) => {

    new WsClientEvented(wsUrl, wsProtocol, {
      onWsOpen: (e, instance) => {
        instance.close();
      },
      onWsClose: (e) => {
        expect(e.code).to.equal(1000);
        done();
      }
    });
  });
});

describe('when sending a message', () => {

  let instance;
  beforeEach(() => {
    instance = new WsClientEvented(wsUrl, wsProtocol, {
      autoOpen: false,
    });
  });

  it('should always require the message type', () => {
    expect(instance.send).to.throw(/Must provide an event type/);
  });

  it('should use a fluent interface', () => {
    expect(instance.send('x', 'y')).to.be.equal(instance);
  });

  it('should send the data over the open socket if it is connected', (done) => {

    instance.open();
    window.setTimeout(() => {
      instance.send('x', 'y');
      done();
    }, 100);
  });

  it('should automatically emit an event for any event type when it is received', (done) => {

    const newInstance = new WsClientEvented(wsUrl, wsProtocol);
    newInstance.emitter.addListener('test-response', (evt, response) => {
      expect(evt).to.be.instanceOf(MessageEvent);
      expect(response).to.deep.equal({
        type: 'test-response',
        payload: {
          my: 'data',
        },
      });

      done();
    });

    newInstance.send('test', {
      my: 'data',
    });
  });
});

describe('when adding or removing listeners', () => {

  let instance;
  beforeEach(() => {
    instance = new WsClientEvented(wsUrl, wsProtocol, {
      autoOpen: false,
      autoReconnect: false,
    });
  });

  describe('when adding listeners via addListener', () => {

    it('should use a fluent interface', () => {
      expect(instance.addListener('test', () => null)).to.be.equal(instance);
    });

    it('should listen for the occured event', (done) => {

      instance.addListener('mymsg-response', (evt, response) => {
        expect(evt).to.be.instanceOf(MessageEvent);
        expect(response).to.deep.equal({
          type: 'mymsg-response',
          payload: {
            data: 'test',
          },
        });
        done();
      });

      instance.open();
      instance.send('mymsg', {
        data: 'test',
      });
    });
  });

  describe('when removing listeners via removeListener', () => {

    it('should use a fluent interface', () => {
      expect(instance.removeListener('test')).to.be.equal(instance);
    });

    it('should remove the listener for the given event', () => {

      instance.addListener('mymsg-response', () => {});
      expect(instance.emitter.listeners('mymsg-response')).to.have.length(1);

      instance.removeListener('mymsg-response');
      expect(instance.emitter.listeners('mymsg-response')).to.have.length(0);
    });
  });
});


cfg/dist.js
var path = require('path');
var webpack = require('webpack');

var config = {
  debug: true,
  entry: {
    index: [ path.join(__dirname, '../lib/index.js') ],
    browser: [ path.join(__dirname, '../lib/browser.js') ]
  },
  output: {
    path: path.join(__dirname, '/../dist'),
    filename: '[name].js'
  },
  cache: false,
  devtool: 'sourcemap',
  plugins: [
    new webpack.optimize.DedupePlugin(),
    new webpack.DefinePlugin({
      'process.env.NODE_ENV': '"production"'
    }),
    new webpack.optimize.UglifyJsPlugin(),
    new webpack.optimize.OccurenceOrderPlugin(),
    new webpack.optimize.AggressiveMergingPlugin(),
    new webpack.NoErrorsPlugin()
  ],
  module: {
    preLoaders: [
      {
        test: /\.js$/,
        include: path.join(__dirname, 'lib'),
        loader: 'eslint-loader'
      }
    ],
    loaders: [
      {
        test: /\.(js)$/,
        loader: 'babel',
        include: path.join(__dirname, '/../lib')
      }
    ]
  }
};

module.exports = config;


karma.conf.js
const webpackCfg = require('./webpack.config');

module.exports = function karmaConfig(config) {
  config.set({
    browsers: ['Chrome'],
    files: [
      'test/bootstrapTests.js',
    ],
    port: 9001,
    captureTimeout: 60000,
    frameworks: [
      'mocha',
      'websocket-server',
    ],
    client: {
      mocha: {},
    },
    singleRun: true,
    reporters: ['mocha', 'coverage'],
    mochaReporter: {
      output: 'autowatch',
      showDiff: true,
    },
    preprocessors: {
      'test/bootstrapTests.js': ['webpack', 'sourcemap'],
    },
    webpack: webpackCfg,
    websocketServer: {
      port: 9999,
      beforeStart: (server) => {
        const CURRENT_PROTOCOL = 'wsclient-evented-protocol';

        server.on('request', (req) => {
          if (req.requestedProtocols.indexOf(CURRENT_PROTOCOL) === -1) {
            req.reject(404, `Wanted protocol is not implemented. Available protocols: ${CURRENT_PROTOCOL}`);
            return;
          }

          const connection = req.accept(CURRENT_PROTOCOL, req.origin);
          connection.on('message', (msg) => {
            if (msg.type === 'utf8') {
              const payload = JSON.parse(msg.utf8Data);

              switch (typeof payload) {
                case 'string':
                  connection.sendUTF(JSON.stringify(`${payload} - response`));
                  break;

                case 'object':
                default:
                  connection.sendUTF(JSON.stringify({
                    type: `${payload.type}-response`,
                    payload: payload.payload,
                  }));
              }
            }
          });
        });
      },
    },
    coverageReporter: {
      dir: 'coverage/',
      watermarks: {
        statements: [85, 95],
        functions: [85, 95],
        branches: [85, 95],
        lines: [85, 95],
      },
      reporters: [
        { type: 'text-summary' },
        {
          type: 'html',
          subdir: 'html',
        },
        {
          type: 'cobertura',
          subdir: 'cobertura',
        },
        {
          type: 'lcovonly',
          subdir: 'lcov',
        },
      ],
    },
  });
};


lib/index.js
/* eslint no-console: 0 */
import { EventEmitter } from 'fbemitter';

/**
 * Default regular emitter
 * @param {MessageEvent} e Original websocket event
 * @param {WsClientEvented} instance
 */
const defaultMessageHandler = (e, instance) => {
  const data = JSON.parse(e.data);
  const responseType = typeof data;

  if (responseType === 'object' && data.type) {
    instance.emitter.emit(data.type, e, data);
  }
};

/**
 * WSClientEvented - WebSocket Client with event emission
 * @author Weblogixx (cs@weblogixx.de)
 *
 * @example let wsClient = new WsClientEvented('ws://localhost:123', 'example-protocol', {
 *   autoOpen: false
 * });
 *
 * Available options:
 * bool autoOpen [default: true] Automatically connect the websocket on initialisation?
 * bool autoReconnect [default: true] Automatically reconnect the websocket if connection is lost?
 * bool debug [default: false] Show debug output?
 * bool forceCloseOnReload [default: false] Send a forced close on browser reload? Only available when using the client in the main thread!
 * int reconnectInterval [default: 1000] Delay in ms for reconnect tries
 * float reconnectDecay [default 1.5] Factor for reconnects
 * int timeoutInterval [default: 2000] Time in ms to wait for connects
 * string binaryType [default: blob] Binary type for websocket transmission
 * string payloadItemName [default: payload] Send key for send items
 * int maxReconnects [default: null] Maximal amount of reconnects. Set to null for inifinite
 * int maxReconnectTimeout [default: 5000] Maximal amount of milliseconds to wait for reconnects
 * int maxSendTries [default: 10] Maximal amount of tries for ws send to fail for the same request
 * function onBeforeWsOpen [default: null] Called before WebSocket connections are established
 * function onWsOpen [default: null] Called when a WebSocket connection becomes ready
 * function onWsMessage [default: defaultMessageHandler] Called on all WebSocket onMessage events. Defaults to emitting via fbemitter
 * function onWsClose [default: null] Called when a WebSocket connection is closed
 * function onWsError [default: null] Called when a WebSocket connection throws errors
 * function onMaxReconnects [default: null] Called when the maximal amount of reconnects is reached
 * function onWsTimeout [default: null] Called when a WebSocket connection times out
 */
export default class WsClientEvented {

  /**
   * Create a new instance of WsClientEvent
   * @param {String} url Url to connect to
   * @param {String|Array} protocols List of protocols [optional]
   * @param {Object} options Options to pass in [optional]
   */
  constructor(url, protocols = '', options = {}) {

    // Instanciate a new event emitter
    this.emitter = new EventEmitter();

    this.url = url;
    this.protocols = protocols && protocols.length > 0 ? protocols : null;

    // How often did the websocket retry to connect?
    this.reconnectAttempts = 0;

    // Set to true to indicate no reconnect should take place
    this.forcedClose = false;

    // The websocket instance
    this.ws = null;

    // List of tries for each send request
    this.sendTries = {};

    // Default settings
    this.settings = {
      autoOpen: true,
      autoReconnect: true,
      debug: false,
      forceCloseOnReload: false,
      reconnectInterval: 1000,
      reconnectDecay: 1.5,
      timeoutInterval: 2000,
      binaryType: 'blob',
      payloadItemName: 'payload',
      maxReconnects: null,
      maxSendTries: 10,
      maxReconnectTimeout: 5000,
      onBeforeWsOpen: null,
      onWsOpen: null,
      onWsMessage: defaultMessageHandler.bind(this),
      onWsClose: null,
      onWsError: null,
      onMaxReconnects: null,
      onWsTimeout: null,
    };

    // Merge settings with defaults
    Object.keys(options).forEach((option) => {
      this.settings[option] = options[option];
    });

    // Attach all active listeners
    ['onBeforeWsOpen', 'onWsOpen', 'onWsMessage', 'onWsClose', 'onWsError', 'onMaxReconnects', 'onWsTimeout'].forEach((userEvt) => {
      if (typeof this.settings[userEvt] === 'function') {
        this.emitter.addListener(userEvt, (evt) => {
          this.settings[userEvt](evt, this);
        });
      }
    });

    if (this.settings.autoOpen) {
      this.open();
    }

    if (this.settings.forceCloseOnReload) {
      self.addEventListener('beforeunload', () => {
        this.close();
      });
    }
  }

  /**
   * Output debug information (only if debug setting enabled)
   * @param  {String} msg Message to output
   */
  debug(msg) {
    if (this.settings.debug) {
      console.warn(msg);
    }
  }

  /**
   * Add a new event listener
   * @param {String} evt Name of the event
   * @param {Function} callback Function to attach
   * @return {WsClientEvented}
   */
  addListener(evt, callback) {
    this.emitter.addListener(evt, callback);
    return this;
  }

  /**
   * Removes all listeners with a given name
   * @param  {String} evt
   * @return {WsClientEvented}
   */
  removeListener(evt) {
    this.emitter.removeAllListeners(evt);
    return this;
  }

  /**
   * Send a new request
   * @param  {String} event Event to send
   * @param  {Object|String} payload Data to send as payload [optional]
   * @return {WsClientEvent}
   */
  send(type, payload = {}) {

    // Skip if type is not given
    if (!type) {
      throw new Error(`Must provide an event type, provided: ${type}`);
    }

    // Only send if the connection is available.
    if (!this.ws) {
      if (this.settings.autoReconnect) {
        this.open();
      } else {
        this.debug(`Tried to send event ${type}, but forced close was set. Please open the connection manually via #open!`);
        return this;
      }
    }

    // Add the item to the send try list
    if (typeof this.sendTries[type] === 'undefined') {
      this.sendTries[type] = 1;
    }

    this.debug(`Trying to send event ${type} for the ${this.sendTries[type]} time...`);

    switch (this.ws.readyState) {

      // If the socket is in connecting state, try it again in a couple of ms
      case WebSocket.CONNECTING:

        // Skip trying to send the request if we reached the maximum
        if (this.sendTries[type] >= this.settings.maxSendTries) {
          this.debug(`WS still in connecting state, giving up after ${this.settings.maxSendTries} tries for event ${type}`);
          return this;
        }

        this.sendTries[type] += 1;
        self.setTimeout(() => {
          this.send(type, payload);
        }, 500);
        break;

      // If the socket is getting closed or is closed, reopen it
      case WebSocket.CLOSING:
      case WebSocket.CLOSED:

        // Skip trying to send the request if we reached the maximum
        if (this.sendTries[type] >= this.settings.maxSendTries) {
          this.debug(`WS still in closing state, giving up after ${this.settings.maxSendTries} tries for event ${type}`);
          return this;
        }

        this.sendTries[type] += 1;
        this.open(true);
        self.setTimeout(() => {
          this.send(type, payload);
        }, 500);
        break;

      // If the websocket is open, just send
      case WebSocket.OPEN:
      default:

        this.ws.send(JSON.stringify({
          type,
          [this.settings.payloadItemName]: payload,
        }));

        // Request went through, reset the try counter
        this.sendTries[type] = 1;
    }

    return this;
  }

  /**
   * Open the websocket
   * @param {Boolean} reconnecting Status of the connection (true if a forced reconnect was done)
   * @return {WsClientEvent}
   */
  open(reconnecting = false) {

    // If the maximal amounts of reconnects is hit, just exit
    if (this.settings.maxReconnects && this.reconnectAttempts >= this.settings.maxReconnects) {
      this.debug('Maximal amount of reconnects reached, giving up! Please call #open by yourself to retry.');
      this.emitter.emit('onMaxReconnects');
      return this;
    }

    // Close the websocket if it is already open
    if (!reconnecting && this.ws) {
      this.close();
    }

    this.emitter.emit('onBeforeWsOpen');

    // Automatically disconnect after a given period of time
    const connectTimeOut = setTimeout(() => {
      this.debug('WS Connection timed out');
      this.emitter.emit('onWsTimeout');
    }, this.settings.timeoutInterval);

    // Create the new websocket instance
    this.ws = new WebSocket(this.url, this.protocols);
    this.ws.binaryType = this.settings.binaryType;

    // Add the needed events
    this.ws.addEventListener('open', (e) => {

      // Clean up the connect timeout after open was fired
      clearTimeout(connectTimeOut);

      this.emitter.emit('onWsOpen', e, this);

      this.debug(`WS connection successfully opened on attemp#${this.reconnectAttempts}.`);

      // We are connected, so reset the reconnect attempts
      this.reconnectAttempts = 0;

      // Reset the send queue
      this.sendTries = {};
    });

    this.ws.addEventListener('message', (e) => {
      this.emitter.emit('onWsMessage', e, this);
    });

    this.ws.addEventListener('close', (e) => {

      clearTimeout(connectTimeOut);

      this.emitter.emit('onWsClose', e, this);

      // Try to reconnect if the close action was not forced via WsClientEvented#close
      if (!this.forcedClose && this.settings.autoReconnect) {

        // Reconnect the websocket automatically
        // @credits: https://github.com/joewalnes/reconnecting-websocket/
        let timeout = this.settings.reconnectInterval * Math.pow(this.settings.reconnectDecay, this.reconnectAttempts);
        if (timeout > this.settings.maxReconnectTimeout) {
          timeout = this.settings.maxReconnectTimeout;
        }
        this.debug(`Auto reconnect required, trying again in ${timeout} ms...`);

        setTimeout(() => {
          this.reconnectAttempts += 1;
          this.open(true);
          this.debug('WS connection timed out, trying to reconnect...');
        }, timeout);
      } else if (this.forcedClose) {
        this.forcedClose = false;
        this.debug('WS forced closing the connection');
      }
    });

    this.ws.addEventListener('error', (e) => {
      this.emitter.emit('onWsError', e, this);
    });

    return this;
  }

  /**
   * Close the websocket if it is open
   * @param  {Number} code The code to send [optional]
   * @param  {String} reason The reason for closing [optional]
   * @return {WsClientEvent}
   */
  close(code = 1000, reason = '') {
    if (this.ws) {
      this.forcedClose = true;
      this.reconnectAttempts = 0;
      this.sendTries = {};
      this.ws.close(code, reason);
      this.ws = null;
    }
    return this;
  }
}
